burn-import/src/onnx/state.rs

```diff
@@ -0,0 +1,246 @@
+use std::{marker::PhantomData, path::Path};
+
+use crate::onnx::{ir::TensorData, op_configuration::linear_config};
+
+use super::{
+    from_onnx::parse_onnx,
+    ir::{self, ArgType, Graph, Node, NodeType},
+    op_configuration::{batch_norm_config, conv2d_config},
+    shape_inference::first_input_dim,
+};
+
+use burn::{
+    module::{Module, Param},
+    nn::{
+        conv::{Conv2dRecord, Conv2dRecordItem},
+        BatchNormRecordItem, LinearRecord, LinearRecordItem,
+    },
+    record::{Record, RecordSettings},
+    tensor::Tensor,
+};
+
+use burn_ndarray::NdArrayBackend;
+
+use serde::{
+    de::Deserializer,
+    ser::{SerializeMap, Serializer},
+    Deserialize, Serialize,
+};
+
+type B = NdArrayBackend<f32>;
+
+pub struct ModelState<RS: RecordSettings> {
+    pub graph: Graph,
+    _record_settings: PhantomData<RS>,
+}
+
+impl<RS: RecordSettings> ModelState<RS> {
+    /// Create a new model state from the onnx file
+    pub fn new<P: AsRef<Path>>(onnx_path: P) -> Self {
+        let graph = parse_onnx(onnx_path.as_ref());
+        Self {
+            graph,
+            _record_settings: PhantomData::default(),
+        }
+    }
+
+    pub fn new_from_graph(graph: Graph) -> Self {
+        Self {
+            graph,
+            _record_settings: PhantomData::default(),
+        }
+    }
+}
+
+impl<RS: RecordSettings> Record for ModelState<RS> {
+    type Item<S: RecordSettings> = ModelState<RS>;
+
+    fn into_item<S: RecordSettings>(self) -> Self::Item<S> {
+        self
+    }
+
+    fn from_item<S: RecordSettings>(_item: Self::Item<S>) -> Self {
+        unimplemented!()
+    }
+}
+
+impl<RS: RecordSettings> Serialize for ModelState<RS> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        // Collect all the stateful nodes
+        let stateful_nodes: Vec<_> = self
+            .graph
+            .nodes
+            .iter()
+            .filter(|node| node.is_stateful)
+            .collect();
+
+        // Create a map to serialize the stateful nodes
+        let mut map = serializer.serialize_map(Some(stateful_nodes.len()))?;
+
+        // Serialize each stateful node into a map entry
+        for node in stateful_nodes.into_iter() {
+            match node.node_type {
+                NodeType::Conv2d => {
+                    let (name, record_item) = conv2d_state::<RS>(node);
+                    map.serialize_entry(&name, &record_item)?;
```

Comment:
Here we are not serializing a vector of nodes rather we convert vectors to a map (node name = key) on the fly. I thought about implement serializer directly on the node itself but it does not reduce the amount of code since you still end up with a match statement for different node types. I agree we should improve the match statement somehow. Maybe pass the whole node to a handler function that mutates  map variable. 

For now, lets put a simple todo and come back later when we add more node implementation because any marginal improvements will still be reimplemented when we have more cases. This has been my experience with this ONNX import. The more I add, my previously refactored code is not sufficient. I am in favor to refactor.

Reply:
Outcome: Added a todo to refactor the matching arms